magpie/alembic/versions/03b54feffe45_ensure_anonymous_group_applied_to_users.py

```diff
@@ -35,17 +60,23 @@ def upgrade():
     context.connection.engine.dialect.supports_sane_multi_rowcount = False
 
     if isinstance(context.connection.engine.dialect, PGDialect):
+        all_user_ids = set(session.execute(sa.select([users.c.id])))
+
         # make sure group exists, then get it
-        init_anonymous(db_session=session)
-        anonym_group = GroupService.by_group_name(constants.get_constant("MAGPIE_ANONYMOUS_GROUP"), db_session=session)
-
-        all_users = BaseService.all(models.User, db_session=session)
-        all_user_group_refs = BaseService.all(models.UserGroup, db_session=session)
-        all_user_group_tups = [(ugr.user_id, ugr.group_id) for ugr in all_user_group_refs]
-        for user in all_users:
-            if (user.id, anonym_group.id) not in all_user_group_tups:
-                user_group = models.UserGroup(user_id=user.id, group_id=anonym_group.id)  # noqa
-                session.add(user_group)
+        anonym_name = get_constant("MAGPIE_ANONYMOUS_GROUP")
+        query = sa.select([groups]).where(groups.c.group_name == anonym_name)
+        anonym_group = session.execute(query).fetchone()
+        if not anonym_group:
+            session.execute(groups.insert().values(group_name=anonym_name, member_count=len(all_user_ids)))
+            anonym_group = session.execute(query).fetchone()
+
+        query = sa.select([users_groups.c.user_id]).where(users_groups.c.group_id == anonym_group.id)
+        users_with_anonym = session.execute(query)
+        users_with_anonym = set(users_with_anonym or [])  # handle if None
+        missing_grp_users = all_user_ids - users_with_anonym
+        missing_grp_users = [{"user_id": usr_id, "group_id": anonym_group.id} for usr_id in missing_grp_users]
+        if missing_grp_users:
+            session.execute(users_groups.insert(), missing_grp_users)
```

Comment:
This comment apply to all alembic patch update. Apart from the discoverable group patch which is new and will be applied, how can we know that the old patches are still working after modifications? Do they are all executed for a new deployment? Are they part of a test procedure?

Reply:
Locally, hard to say, but during Travis-CI execution, it always starts from scratch.

Reply:
I agree backups have to be done and migrating bd is not without risks. At least tests cover the code that is modified. My fear was that you updated old upgrade fct to use new utility fct without possibility to test them. But obviously you had tests for that! :)